### PR TITLE
Add Event Logging to Add Command

### PR DESCRIPTION
**Pull request by frrist**
License: MIT
Signed-off-by: ForrestWeston <forrest@protocol.ai>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
We really need benchmarks...

**Comment by ForrestWeston** [comment]
@Stebalien 
Before Changes (bbdbd0ac6fdcf1f4fdfe84869d5bde83b9e9e79f):
* 0.10user 0.07system 0:04.57elapsed 3%CPU (0avgtext+0avgdata 29340maxresident)k
0inputs+0outputs (0major+2198minor)pagefaults 0swaps

After Changes (d84e5a297afdba15b7dd563ac7cdc3094f30e62c):
* 0.09user 0.08system 0:04.83elapsed 3%CPU (0avgtext+0avgdata 29112maxresident)k
0inputs+0outputs (0major+2204minor)pagefaults 0swaps

**Comment by whyrusleeping** [comment]
Cool, doesnt appear to be any noticeable difference there.

**Comment by whyrusleeping** [comment]
Could you post an image here of the tracing output that this generates?

**Comment by ForrestWeston** [comment]
Sorry it got cut off a bit, this on is from the above benchmarks.
add-go-ipfs

This is one from my `bin/` dir
https://ipfs.io/ipfs/QmUMJCmHRQ2svFjWjXfiC5y5XCHG2qj4TvA4Jbz7XKFvaE/Selection_008.png

File: core/coreunix/add.go [hunk 1]
```diff
@@ -153,112 +153,148 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
 }
 
 func (adder *Adder) RootNode() (node.Node, error) {
```
Comment (ForrestWeston) [inline core/coreunix/add.go]: Does this make sense as well? It feels cleaner, it compiles and the tests pass.
```diff
diff --git a/core/coreunix/add.go b/core/coreunix/add.go
index 516b76d44..090ecee27 100644
--- a/core/coreunix/add.go
+++ b/core/coreunix/add.go
@@ -152,40 +152,42 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
        return balanced.BalancedLayout(params.New(chnk))
 }
 
-func (adder *Adder) RootNode() (node.Node, error) {
+func (adder *Adder) RootNode() (root node.Node, err error) {
        eip := log.EventBegin(adder.ctx, "RootNode")
-       defer eip.Done()
+       defer func() {
+               if root != nil {
+                       eip.Append(root.Cid())
+               }
+               if err != nil {
+                       eip.SetError(err)
+               }
+               eip.Done()
+       }()
        // for memoizing
-       if adder.root != nil {
-               eip.Append(adder.root.Cid())
-               return adder.root, nil
+       root = adder.root
+       if root != nil {
+               return root, nil
        }
 
        mr, err := adder.mfsRoot()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
-       root, err := mr.GetValue().GetNode()
+       root, err = mr.GetValue().GetNode()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
 
        // if not wrapping, AND one root file, use that hash as root.
        if !adder.Wrap && len(root.Links()) == 1 {
-               nd, err := root.Links()[0].GetNode(adder.ctx, adder.dagService)
+               root, err = root.Links()[0].GetNode(adder.ctx, adder.dagService)
                if err != nil {
-                       eip.SetError(err)
                        return nil, err
                }
-
-               root = nd
        }
 
        adder.root = root
-       eip.Append(root.Cid())
-       return root, err
+       return
 }
```

Comment (Stebalien) [reply core/coreunix/add.go]: I'm fine with naming the return values for the purpose of accessing them from within the `defer` statement. However, I wouldn't use the implicit return feature (IMO, it was a mistake). How about:

```
diff --git a/core/coreunix/add.go b/core/coreunix/add.go
index 516b76d44..090ecee27 100644
--- a/core/coreunix/add.go
+++ b/core/coreunix/add.go
@@ -152,37 +152,38 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
        return balanced.BalancedLayout(params.New(chnk))
 }
 
-func (adder *Adder) RootNode() (node.Node, error) {
+func (adder *Adder) RootNode() (root node.Node, err error) {
        eip := log.EventBegin(adder.ctx, "RootNode")
-       defer eip.Done()
+       defer func() {
+               if root != nil {
+                       eip.Append(root.Cid())
+               }
+               if err != nil {
+                       eip.SetError(err)
+               }
+               eip.Done()
+       }()
        // for memoizing
-       if adder.root != nil {
-               eip.Append(adder.root.Cid())
-               return adder.root, nil
+       if addr.root != nil {
+               return addr.root, nil
        }
 
        mr, err := adder.mfsRoot()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
-       root, err := mr.GetValue().GetNode()
+       root, err = mr.GetValue().GetNode()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
 
        // if not wrapping, AND one root file, use that hash as root.
        if !adder.Wrap && len(root.Links()) == 1 {
-               nd, err := root.Links()[0].GetNode(adder.ctx, adder.dagService)
+               root, err = root.Links()[0].GetNode(adder.ctx, adder.dagService)
                if err != nil {
-                       eip.SetError(err)
                        return nil, err
                }
-
-               root = nd
        }
 
        adder.root = root
-       eip.Append(root.Cid())
        return root, err
 }
```

Comment (whyrusleeping) [reply core/coreunix/add.go]: agreed, the implicit return thing was a terrible choice by the go team

Comment (ForrestWeston) [reply core/coreunix/add.go]: Alright, I can make these changes on the rest of the methods here too then. I am not sure if I should make that a separate commit or rebase. What would you prefer?

Comment (whyrusleeping) [reply core/coreunix/add.go]: either way, i'm not picky. Though I do appreciate smaller numbers of commits

File: blockservice/blockservice.go [hunk 2]
```diff
@@ -199,12 +206,14 @@ func getBlock(ctx context.Context, c *cid.Cid, bs blockstore.Blockstore, f excha
...
 // NB: No guarantees are made about order.
 func (s *blockService) GetBlocks(ctx context.Context, ks []*cid.Cid) <-chan blocks.Block {
+	ctx = log.EventBeginInContext(ctx, "BlockService.GetBlocks")
 	return getBlocks(ctx, ks, s.blockstore, s.exchange)
 }
 
 func getBlocks(ctx context.Context, ks []*cid.Cid, bs blockstore.Blockstore, f exchange.Fetcher) <-chan blocks.Block {
 	out := make(chan blocks.Block)
 	go func() {
+		defer logging.MaybeFinishEvent(ctx)
```
Comment (ForrestWeston) [inline blockservice/blockservice.go]: [This PR](https://github.com/ipfs/go-log/pull/8) needs to go first for this functionality.

**Comment by frrist** [comment]
Note: blocked until changes in go-log are bubbled up

**Comment by Stebalien** [comment]
> Note: blocked until changes in go-log are bubbled up - CI will fail till then

Now that everything else has been unblocked, what exact changes do you still need (if any)?

**Comment by frrist** [comment]
@Stebalien we will need to gx publish the changes in go-log and update the version that go-ipfs and other packages use (currently using 1.2.0). 

The changes in go-log include instrumenting it with opentracing, as well as adding the DoneWithErr method and MaybeFinishEvent method (basically all the changes made to go-log in prs 8 - 16). I can start that process now.

**Comment by frrist** [comment]
Closing, will reopen when this becomes a priority -- I currently lack the bandwidth to push this forward.

**Pull request closed without merging**